Icon widget extension: service calls, state update delay

JakubMacoun opened the pull request:
This extends the functionality of _icon_ widget by the ability to make service calls and adding delay to react to state update.
Service calls are possible by using _post_service_active_ parameter for each state. On icon click, specified service is run.
State update delay is specified by _update_delay_ parameter which is in seconds. If specified, on monitored entity state update, rerendering of the icon is delayed by a specified amount of seconds. This is useful when the monitored entity state takes some time to change properly and icon blinks briefly.

Usage example:
```
ac_control:
    widget_type: icon
    entity: sensor.ac_db_control_option
    icons:
        "turn_on_temporarily":
            post_service_active:
                service: sequence/run
                entity_id: sequence.ac_on_temporarily
            icon: mdi-air-conditioner
        "turn_on":
            post_service_active:
                service: homeassistant/turn_on
                entity_id: script.turn_on_ac
            icon: mdi-air-conditioner
        "turn_off":
            post_service_active:
                service: homeassistant/turn_on
                entity_id: script.turn_off_ac
            icon: mdi-air-conditioner
            style: "color: aquamarine;"
    state_text: 1
    update_delay: 1
    state_map:
        "turn_on_temporarily": Turn AC on temporarily
        "turn_on": Turn AC on
        "turn_off": Turn AC off
```

Odianosen25 commented:
@JakubMacoun,

Nice update and good you took @ReneTode's advice. But to make things easier instead of using `script_entity`, you could just have it as `script`, and for the sequence, the user doesn't have to have the sequence specified in the widget. The `sequence` should be defined in AD, and just like the `script`, the `sequence` is used, just similar to the sequence widget if it makes sense.

Regards

JakubMacoun commented:
> @JakubMacoun,
> 
> Nice update and good you took @ReneTode's advice. But to make things easier instead of using `script_entity`, you could just have it as `script`, and for the sequence, the user doesn't have to have the sequence specified in the widget. The `sequence` should be defined in AD, and just like the `script`, the `sequence` is used, just similar to the sequence widget if it makes sense.
> 
> Regards

Thank you for your suggestions. I have renamed _script_entity_ to _script_, you're right, it's easier this way.
About the sequences - I have updated my example and replaced the sequence definition by its name. I'm just not sure if `self.call_service()` can handle the sequence name like this. Documentation mentions `self.run_sequence()` but that's not defined in WidgetBase.

Odianosen25 commented:
Hello @JakubMacoun,

Yes it is possible to use `call_service` for sequences, it is [here](https://appdaemon.readthedocs.io/en/latest/AD_API_REFERENCE.html?highlight=services#services) using `self.call_service("sequence/run", sequence=sequence, namespace="rules")`. You have the be careful abt the `namespace` as the sequence service call only runs within the `rules` namespace 

Regards

JakubMacoun commented:
> Hello @JakubMacoun,
> 
> Yes it is possible to use `call_service` for sequences, it is [here](https://appdaemon.readthedocs.io/en/latest/AD_API_REFERENCE.html?highlight=services#services) using `self.call_service("sequence/run", sequence=sequence, namespace="rules")`. You have the be careful abt the `namespace` as the sequence service call only runs within the `rules` namespace
> 
> Regards

Thank you very much for your help. I have updated usage of `self.call_service()` to be able to work with sequences.

ReneTode commented:
a better way would be the option to offer a service like other widgets do.
```
ac_control:
    widget_type: icon
    entity: sensor.ac_db_control_option
    icons:
        "turn_on_temporarily":
            post_service_active:
                service: homeassistant/turn_on
                entity_id: sequence.ac_on_temporarily
            icon: mdi-air-conditioner
        "turn_on":
            post_service_active:
                service: homeassistant/turn_on
                entity_id: script.ac_turn_on
            icon: mdi-air-conditioner
        "turn_off":
            post_service_active:
                service: homeassistant/turn_on
                entity_id: script.ac_turn_off
            icon: mdi-air-conditioner
            style: "color: aquamarine;"
    state_text: 1
    update_delay: 1
    state_map:
        "turn_on_temporarily": Turn AC on temporarily
        "turn_on": Turn AC on
        "turn_off": Turn AC off
```
that way all possible services are available, not just script or sequence.
an example use for that would be that you can use icons for an input_select.

JakubMacoun commented:
> a better way would be the option to offer a service like other widgets do.
> 
> ```
> ac_control:
>     widget_type: icon
>     entity: sensor.ac_db_control_option
>     icons:
>         "turn_on_temporarily":
>             post_service_active:
>                 service: homeassistant/turn_on
>                 entity_id: sequence.ac_on_temporarily
>             icon: mdi-air-conditioner
>         "turn_on":
>             post_service_active:
>                 service: homeassistant/turn_on
>                 entity_id: script.ac_turn_on
>             icon: mdi-air-conditioner
>         "turn_off":
>             post_service_active:
>                 service: homeassistant/turn_on
>                 entity_id: script.ac_turn_off
>             icon: mdi-air-conditioner
>             style: "color: aquamarine;"
>     state_text: 1
>     update_delay: 1
>     state_map:
>         "turn_on_temporarily": Turn AC on temporarily
>         "turn_on": Turn AC on
>         "turn_off": Turn AC off
> ```
> 
> that way all possible services are available, not just script or sequence.
> an example use for that would be that you can use icons for an input_select.

Thank you for your comment, @ReneTode . I have added an option to call any service and I have also kept options to call just script or sequence, to make things easier. I have updated my example in first comment.

Odianosen25 commented:
@JakubMacoun,

Don’t know if you saw my comment, but the sequence service call should use “entity_id” and not “sequence” as I said earlier. So you will need to change that.

Regards

JakubMacoun commented:
> @JakubMacoun,
> 
> Don’t know if you saw my comment, but the sequence service call should use “entity_id” and not “sequence” as I said earlier. So you will need to change that.
> 
> Regards

@Odianosen25 ,
I'm not sure if I understand, in [this comment](https://github.com/AppDaemon/appdaemon/pull/1019#issuecomment-670935935) you mentioned using sequence. So if I understand you right, this is correct implementation?

```
self.post_service = {
    service: "sequence/run",
    entity_id: "sequence." + data.sequence,
    namespace: "rules"
}
```

Odianosen25 commented:
Yes you right

ReneTode commented:
> Thank you for your suggestion, @ReneTode . I have added an option to call any service and I have also kept options to call just script or sequence, to make things easier. I have updated my example in the first comment.

if i would have anything to say about it (which i dont have anymore) i would suggest to delete the script and sequence options.
if that is added the next thing is that people want turn on and turn off as easy option, and then more.
which would be actually more logical then sequence or script.
and before you know it you are adding 100 different things as easy options, and still it wouldnt be complete, but the docs would get so confusing that noone understands what is possible and what not.
adding services gives all options. keeping script and sequence on the side only saves a few characters of yaml.

and dont forget that to get the PR complete, also the docs need to change. all options need to be described in the new docs.

JakubMacoun commented:
> > Thank you for your suggestion, @ReneTode . I have added an option to call any service and I have also kept options to call just script or sequence, to make things easier. I have updated my example in the first comment.
> 
> if i would have anything to say about it (which i dont have anymore) i would suggest to delete the script and sequence options.
> if that is added the next thing is that people want turn on and turn off as easy option, and then more.
> which would be actually more logical then sequence or script.
> and before you know it you are adding 100 different things as easy options, and still it wouldnt be complete, but the docs would get so confusing that noone understands what is possible and what not.
> adding services gives all options. keeping script and sequence on the side only saves a few characters of yaml.
> 
> and dont forget that to get the PR complete, also the docs need to change. all options need to be described in the new docs.

You're right, too many simplifications are not a good idea. I've kept _service_call_ option only and updated docs.

ReneTode commented:
seems great.
but i am a nitpicker ;) 

i would suggest to keep the same naming that is used in all other widgets.
so not call_service but

post_service_...:

in this case i think it should be post_service_active, because it happens after the icon was active.
see all other yaml files.

JakubMacoun commented:
> seems great.
> but i am a nitpicker ;)
> 
> i would suggest to keep the same naming that is used in all other widgets.
> so not call_service but
> 
> post_service_...:
> 
> in this case i think it should be post_service_active, because it happens after the icon was active.
> see all other yaml files.

Makes sense, done.

ReneTode commented:
now all you need to do is change the docs you made ;)

JakubMacoun commented:
> now all you need to do is change the docs you made ;)

I'm afraid I don't understand. It's changed, _service_call_ is rewritten to _post_service_active_.

JakubMacoun commented:
> > now all you need to do is change the docs you made ;)
> 
> I'm afraid I don't understand. It's changed, _service_call_ is rewritten to _post_service_active_.

I've written some more details to the post_service_active description, so it's easier to understand. Is that what you meant?

ReneTode commented:
sorry, i am bad with reading commits ;)
i missed that you did change it in the docs.

but i must say that the last change you made to the docs explains it a bit better.

so from me you got a big thumbs up.

i think ill test it in a few days, but i cant see anything against this. (but remember im just an ex dev, and no active teammember anymore)

but from my point of view this is a good extra feature for dashboard.

Odianosen25 commented:
@ReneTode, can you pls confirm this so I can merge and close it?

regards

ReneTode commented:
> but from my point of view this is a good extra feature for dashboard.

this was my last remark, so i guess its good ;)
